### PR TITLE
Use get_attnum to find the attribute number of target entry

### PR DESCRIPTION
**Pull request by SaitTalhaNisanci**
This is mostly done for PG14 support.

With Postgres 14, with UPDATE queries the targetEntry->resno no longer points to the attribute number hence some of our logic became broken.

We can use `get_attnum` to get attribute number correctly, so we switch to this approach.

### CODE REVIEW

File: src/test/regress/sql/modification_correctness.sql [hunk 1]
```diff
@@ -0,0 +1,51 @@
+CREATE SCHEMA modification_correctness;
+
+CREATE TABLE test(a int, b int, c int unique, d int, e int);
```
Comment (onderkalaci) [inline src/test/regress/sql/modification_correctness.sql:3]: set search path?

File: src/backend/distributed/planner/multi_router_planner.c [hunk 2]
```diff
@@ -634,24 +635,42 @@ ModifyPartialQuerySupported(Query *queryTree, bool multiShardQuery,
...
 			TargetEntry *targetEntry = (TargetEntry *) lfirst(targetEntryCell);
+
+			/* skip resjunk entries: UPDATE adds some for ctid, etc. */
+			if (targetEntry->resjunk)
+			{
+				continue;
+			}
+
 			bool targetEntryPartitionColumn = false;
+			AttrNumber attNumber = 0;
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_router_planner.c:646]: minor: use `InvalidAttrNumber`

Also, `attNunber` is way too generic here making it hard to follow. Can you rename throughout the PR as `targetColumnAttrNumber` or such?

File: src/backend/distributed/planner/multi_router_planner.c [hunk 3]
```diff
@@ -1517,12 +1545,13 @@ MasterIrreducibleExpressionFunctionChecker(Oid func_id, void *context)
  * tree, or the target entry sets a different column.
  */
 static bool
-TargetEntryChangesValue(TargetEntry *targetEntry, Var *column, FromExpr *joinTree)
+TargetEntryChangesValue(TargetEntry *targetEntry, Var *column, FromExpr *joinTree,
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_router_planner.c:1548]: I think `TargetEntryChangesValue` requires some more comments. We pass `targetEntry` and `targetColumnAttrNumber`. And, I cannot clearly follow why do we have these two cases (e.g., why not the second check is not enough for all cases?):
```C
	if (targetColumnAttrNumber != column->varattno)
	{
		/* target entry of the form SET some_other_col = <x> */
		isColumnValueChanged = false;
	}
	else if (IsA(setExpr, Var))
	{
		Var *newValue = (Var *) setExpr;
		if (newValue->varattno == column->varattno)
		{
			/* target entry of the form SET col = table.col */
			isColumnValueChanged = false;
		}
	}
```

File: src/test/regress/expected/modification_correctness.out [hunk 4]
```diff
@@ -0,0 +1,51 @@
+CREATE SCHEMA modification_correctness;
```
Comment (onderkalaci) [inline src/test/regress/expected/modification_correctness.out:1]: maybe some tests with `citus.enable_fast_path_router_planner to false` ?

File: src/test/regress/expected/modification_correctness.out [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+INSERT INTO test (c,a) VALUES(3,4) ON CONFLICT(c) DO UPDATE SET c=EXCLUDED.c, d = EXCLUDED.c;
+INSERT INTO test (c,a) VALUES(3,4) ON CONFLICT(c) DO UPDATE SET c=EXCLUDED.c, d = EXCLUDED.c, e = EXCLUDED.c;
+INSERT INTO test (c,a) VALUES(3,4) ON CONFLICT(c) DO UPDATE SET c=EXCLUDED.c, d = EXCLUDED.c, e = 7;
+PREPARE foo(int,int) AS INSERT INTO test (c,a) VALUES($1,$2) ON CONFLICT(c) DO UPDATE SET c=EXCLUDED.c, d = EXCLUDED.c, e = $1;
+EXECUTE foo(1,2);
+EXECUTE foo(1,2);
+EXECUTE foo(1,2);
+EXECUTE foo(1,2);
+EXECUTE foo(1,2);
+EXECUTE foo(1,2);
```
Comment (onderkalaci) [inline src/test/regress/expected/modification_correctness.out:43]: I think 7th execution is also required.

File: src/test/regress/expected/modification_correctness.out [hunk 6]
```diff
@@ -0,0 +1,51 @@
+CREATE SCHEMA modification_correctness;
+CREATE TABLE test(a int, b int, c int unique, d int, e int);
+SELECT create_distributed_table('test', 'c');
+ create_distributed_table
+---------------------------------------------------------------------
+
+(1 row)
+
+ALTER TABLE test DROP column b;
```
Comment (onderkalaci) [inline src/test/regress/expected/modification_correctness.out:9]: maybe also drop 1 another column before `create distributed table` as well, we had weird edge cases on other code paths.

File: src/test/regress/expected/modification_correctness.out [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+UPDATE test SET a = 5, c = 5;
+ERROR:  modifying the partition value of rows is not allowed
+UPDATE test SET a = 5, c = d, d =3;
+ERROR:  modifying the partition value of rows is not allowed
+UPDATE test SET c = d, d =3;
+ERROR:  modifying the partition value of rows is not allowed
+UPDATE test SET d=c, c = d;
+ERROR:  modifying the partition value of rows is not allowed
+UPDATE test SET e = c, d = 3;
+UPDATE test SET c = c;
```
Comment (onderkalaci) [inline src/test/regress/expected/modification_correctness.out:19]: maybe some basic tests with FROM clause as well? like
```SQL
UPDATE test SET a = 5,d  = 2, c = 5 FROM (SELECT * FROM test LIMIT 10) t2 WHERE t2.d = test.c  and test.c = 5;

UPDATE test SET a = 5,d  = 2, c = 5 FROM (SELECT * FROM test LIMIT 10) t2 WHERE t2.d = test.c  and test.c = 6;

```

**Review by onderkalaci — COMMENTED** [review]
I think in general our checks for unsupported modifications can be simplified.

It is quite hard to follow what are the rules at the moment. So, let's not make things worse and at least add some more comments with these new changes and improve variable names?

Other than that, it looks its working as expected.

File: src/backend/distributed/planner/multi_router_planner.c (hunk 3 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/planner/multi_router_planner.c:1548]: You mean why is 

```
		if (newValue->varattno == column->varattno)
		{
			/* target entry of the form SET col = table.col */
			isColumnValueChanged = false;
		}
```

not enough?

If that is the question, that check is seeing if we have something like `UPDATE test SET c=c`, in which case we count that as `column value not changed`.

File: src/backend/distributed/planner/multi_router_planner.c [hunk 8]
```diff
@@ -634,23 +634,37 @@ ModifyPartialQuerySupported(Query *queryTree, bool multiShardQuery,
...
 			}
-			else if (targetEntry->resno == partitionColumn->varattno)
+			else
 			{
-				targetEntryPartitionColumn = true;
+				if (commandType == CMD_UPDATE)
+				{
+					if (targetEntry->resname)
+					{
+						targetColumnAttrNumber = get_attnum(resultRelationId,
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_router_planner.c:658]: maybe add a single line comment explaining that it is not possible to give an alias to `UPDATE .. SET column `

**Review by onderkalaci — APPROVED** [review]
